File: bprolog/least_diff.pl
/*

  Least diff problem in B-Prolog.

  The model solves the following problem:
 
  What is the smallest difference between two numbers X - Y
  if you must use all the digits (0..9) exactly once, i.e.
  Minimize the difference 
    ABCDE - FGHIJ


  Model created by Hakan Kjellerstrand, hakank@gmail.com
  See also my B-Prolog page: http://www.hakank.org/bprolog/

*/


%
% Reporting both time and backtracks
%
time2(Goal):-
        cputime(Start),
        statistics(backtracks, Backtracks1),
        call(Goal),
        statistics(backtracks, Backtracks2),
        cputime(End),
        T is (End-Start)/1000,
        Backtracks is Backtracks2 - Backtracks1,
        format('CPU time ~w seconds. Backtracks: ~d\n', [T, Backtracks]).

go :-
        time2(least_diff([L,X,Y,Diff], Backtracks)),
        writeln(L),
        format('~d - ~d = ~d\n',[X,Y,Diff]),
        format('~d backtracks\n',[Backtracks]).

go2 :-
        time2(least_diff2([L,X,Y,Diff], Backtracks)),
        writeln(L),
        format('~d - ~d = ~d\n',[X,Y,Diff]),
        format('~d backtracks\n',[Backtracks]).


%
% A small benchmark:
% least_diff2/2 is slightly faster than least_diff/2. 
% For N=1000
%  least_diff : 10.916s
%  least_diff2: 10.009s
%
go3 :-
        N = 1000,
        writeln(first_diff),
        time(
        foreach(_ in 1..N,
                [L,X,Y,Diff, Backtracks],
                least_diff([L,X,Y,Diff],Backtracks))
             ),

        writeln(first_diff2),
        time(
        foreach(_ in 1..N,
                [L,X,Y,Diff, Backtracks],
                least_diff2([L,X,Y,Diff],Backtracks))
             ).



%
% Using scalar_product/4.
%
least_diff([L,X,Y,Diff], Backtracks) :-
        statistics(backtracks, Backtracks1),
        L = [A,B,C,D,E,F,G,H,I,J],
        L :: 0..9,
        length(L, N),
        M is N // 2,

        alldifferent(L),
        % Since 10**0 yield 1.0 instead of 1, we have to use integer/1.
        Base @= [T : I in 1..M, [T], T is integer(10**(M-I))],
        scalar_product(Base,[A,B,C,D,E],#=,X),
        scalar_product(Base,[F,G,H,I,J],#=,Y),
        Diff #= X - Y,
        Diff #> 0,

        minof(labeling([ff,down],L), Diff),

        statistics(backtracks, Backtracks2),
        Backtracks is Backtracks2 - Backtracks1.


%
% Different approach in the arithmetics
%
least_diff2([L,X,Y,Diff], Backtracks) :-
        statistics(backtracks, Backtracks1),
        L = [A,B,C,D,E,F,G,H,I,J],
        L :: 0..9,

        alldifferent(L),
        X #= 10000*A + 1000*B + 100*C + 10*D + E,
        Y #= 10000*F + 1000*G + 100*H + 10*I + J,
        Diff #= X - Y,
        Diff #> 0,

        minof(labeling([ff,down],L), Diff),

        statistics(backtracks, Backtracks2),
        Backtracks is Backtracks2 - Backtracks1.
        

% sat_solve
% (if it's a solution it's > 5minutes)
least_diff3(L,X,Y,Diff) :-
        L = [A,B,C,D,E,F,G,H,I,J],
        L :: 0..9,

        $alldifferent(L),
        X $= 10000*A + 1000*B + 100*C + 10*D + E,
        Y $= 10000*F + 1000*G + 100*H + 10*I + J,
        Diff $= X - Y,
        Diff $> 0,

        sat_solve([min(Diff)], L).
        

%
% Decomposition of alldifferent for ip_solve
%
alldifferent_mip(L) :-
        length(L, Len),
        foreach(I in 1..Len, J in I+1..Len, L[I] $\= L[J]).


% ip_solve
least_diff4(L,X,Y,Diff) :-
        L = [A,B,C,D,E,F,G,H,I,J],
        L :: 0..9,
        lp_domain(L,0,9),
        lp_integers(L),

        foreach(LL in L, lp_domain(LL,0,9)),
        foreach(LL in L, lp_integer(LL)),

        alldifferent_mip(L),
        X $= 10000*A + 1000*B + 100*C + 10*D + E,
        Y $= 10000*F + 1000*G + 100*H + 10*I + J,
        Diff $= X - Y,
        Diff $> 0,

        % ip_solve([dump, min(Diff)], L).
        ip_solve([min(Diff)], L).
        

File: z3/least_diff.py
#!/usr/bin/python -u
# -*- coding: latin-1 -*-
# 
# Least diff problem in Z3
#
# The model solves the following problem:
# 
# What is the smallest difference between two numbers X - Y
# if you must use all the digits (0..9) exactly once, i.e.
# Minimize the difference 
#   ABCDE - FGHIJ
#
# Also, this model compares between using Optimize() and Solver()
# and we see that using different ordering of the constraints yield
# different intermediate solutions for this optimization problem.
# 
#
# This z3 model was written by Hakan Kjellerstrand (hakank@gmail.com)
# See also my z3 page: http://hakank.org/z3/
# 
#
import time
from z3_utils_hakank import *

#
# This version use Optimize() where we only see the final optimal solution.
# 
def least_diff_optimize():
    print("Using Optimize()")
    
    sol = Optimize()

    a,b,c,d,e = Ints("a b c d e")
    f,g,h,i,j = Ints("f g h i j")
    diff = Int("diff")

    t = [a,b,c,d,e, f,g,h,i,j]
    for x in t:
        sol.add(x >= 0)
        sol.add(x <= 9)

    sol.add(Distinct(t))
    sol.add(diff == (10000*a+1000*b+100*c+10*d+e)-(10000*f+1000*g+100*h+10*i+j))
    sol.add(diff >= 0)

    sol.minimize(diff)
    if sol.check() == sat:
        mod = sol.model()
        print("diff=",mod.evaluate(diff))
        print(evalArray(mod,t)) # [mod.evaluate(x) for x in t]
    else:
        print("No solution")

#
# This version use Solver and we manually ensure optimality
# and we see the how the value diff evolves
#
def least_diff_solver(type=1):
    print("Using Solver()")
    # sol = Solver()         # 0.046s for type 1 and  0.071s for type 2
    sol = SolverFor("QF_FD") # 0.0038s for type 1 and 0.0035s for type 2

    a,b,c,d,e = Ints("a b c d e")
    f,g,h,i,j = Ints("f g h i j")
    diff = Int("diff")

    t = [a,b,c,d,e, f,g,h,i,j]
    for x in t:
        sol.add(x >= 0)
        sol.add(x <= 9)

    # The order of the constraints (might) matter:
    # For Solver() type 1 is a little faster
    # For QF_FD type 2 tends to be faster
    if type == 1:
        sol.add(Distinct(t))
        sol.add(diff == (10000*a+1000*b+100*c+10*d+e)-(10000*f+1000*g+100*h+10*i+j))
        sol.add(diff >= 0)
    else:
        # .. than this, where Distinct is last
        sol.add(diff == (10000*a+1000*b+100*c+10*d+e)-(10000*f+1000*g+100*h+10*i+j))
        sol.add(diff >= 0)
        sol.add(Distinct(t))

    while sol.check() == sat:
        mod = sol.model()
        print("diff=",mod.evaluate(diff))
        print(evalArray(mod,t)) # [mod.evaluate(x) for x in t]
        getLessSolution(sol,mod, diff)

t0 = time.time()
least_diff_optimize()
t1 = time.time()
print("Time:", t1-t0)
print()

least_diff_solver(1)
t2 = time.time()
print("Time:", t2-t1)
print()

least_diff_solver(2)
t3 = time.time()
print("Time:", t3-t2)
print()
